### DISCUSSION TITLE
How to select a Ligand/Sugar with a Glycam nomenclature?

### DISCUSSION THREAD

**Discussion by Noah603** [MDAnalysis/mdanalysis#4562] [Q&A] [answered]
Hey guys!

I have performed an MD sim with my Protein and a trisaccharide (2'FL) and penta- and hexa-saccharides. The Ligand(s) are defined in the prmtop file by the Glycam nomenclature: for example for 2'FL: ROH 4GA 2LB 0fA. 
Is there a nice way to just call them collectivly? Or do I always call them individually, like I am currently doing. I tryed to change in the prmtop file the individuall names to just LIG and use that. That kinda works but is not really pleasing.

Best greetings, Noah

**Comment by hmacdope** [comment, marked as answer]
Hi Noah, 

Can you give me an example (in code) of how you are running your selection? 

Cheers, 

Hugo

**Reply by Noah603** [reply]
Sure thing: 
```python
u2 = mda.Universe("noWATnoIONS.prmtop", "21_03_24_LNH_5IYR.crd", format="TRJ")
ligand_selection2 = u2.select_atoms("resname ROH or resname 4GA or resname 2LB or resname 0FA")
```
After that I used prolif functions to create a fingerprint, out of which I created barcode plots. 

To create RMSD plots for the ligand, I used methods from [this tutorial](https://projects.volkamerlab.org/teachopencadd/talktorials/T020_md_analysis.html):

```python
def rmsd_for_atomgroups2(universe, selection1, selection2=None, selection3=None, selection4=None, selection5=None, selection6=None):
    """Calulate the RMSD for selected atom groups.

    Parameters
    ----------
    universe: MDAnalysis.core.universe.Universe
        MDAnalysis universe.
    selection1: str
        Selection string for main atom group, also used during alignment.
    selection2: list of str, optional
        Selection strings for additional atom groups.

    Returns
    -------
    rmsd_df: pandas.core.frame.DataFrame
        DataFrame containing RMSD of the selected atom groups over time.
    """

    universe.trajectory[0]
    ref = universe
    rmsd_analysis = rms.RMSD(universe, ref, select=selection1, groupselections=selection2)
    rmsd_analysis.run()
    columns = [selection1, *selection2] if selection2 else [selection1]
    rmsd_df = pd.DataFrame(np.round(rmsd_analysis.results.rmsd[:, 2:], 2), columns=columns)
    rmsd_df.index.name = "frame"
    return rmsd_df
```
```python
ligand_name ='4GA'
ligand_name2="2LB"
ligand_name3='0FA'
rmsd = rmsd_for_atomgroups2(u, "backbone", ["protein", f"resname {ligand_name}",f"resname {ligand_name2}", f"resname {ligand_name3}"])
rmsd.plot(title="RMSD of protein and ligand")
plt.ylabel("RMSD (Å)");
```

This will give me out a plot in which one can see each RMSD of the given monomers. This can be nice (especially with more compelx sugars), but it would be also nice to see the overall RMSD of the ligand, since is less complex.
I have also tried analysing the H-Bond distances, but the calling gets difficult since the ligand is not treated as one ligand, but rather as multiple ones. For this reason I would also like to call the ligand as a whole.

Hope this helps!

Best, Noah

**Reply by hmacdope** [reply]
If your ligands have a consistent resname You should be able to do something like  the below, apologies havent tested and there may be a neater way but see if this works for you. 

```python
ag = u.select_atoms("resname  MYLIG")
resids = set(ag.resids)

for r in resids:
    ag_local = u.select_atoms(f"resid {r}")
    # do something with ag_local
```

**Reply by Noah603** [reply]
Thank you for the reply!
Do you propose to change the names of this one ligand (ROH 4GA 2LB 0fA) to LIG and go from there?

**Reply by hmacdope** [reply]
Are those the ligand names or the atom names? They look like atom names to me. Would you mind attatching the file?

**Reply by Noah603** [reply]
So, since I work with Oligosaccharides, each monomer has a specifc nomenclature from Glycam ([see here](https://glycam.org/docs/forcefield/glycam-naming-2/index.html)). Hence this wierd looking names. Attached is the prmtop file (hope zipping worked fine). At line 7100 the protein sequence starts and ends with the glycam ligand nomenclature.
[noWATnoIONS.input.prmtop.zip](https://github.com/MDAnalysis/mdanalysis/files/14944470/noWATnoIONS.input.prmtop.zip)

**Reply by hmacdope** [reply]
I might be missing what you are trying to do but  assuming you want to find the RMSD of your whole ligand **with respect to itself**, ** the following should hopefully work. @orbeckst can possibly double check I have things right here? 

```python
import MDAnalysis as mda 
from MDAnalysis.analysis import align


u = mda.Universe(...)
select_all = "protein or resname ROH or resname 4GA or resname 2LB or resname 0FA"
select_ligand = "resname ROH or resname 4GA or resname 2LB or resname 0FA"

# first align to remove translation and rotation from the calculation 
alignment = align.AlignTraj(
    mobile=u, reference=u, select=select_all, in_memory=True
)
alignment.run()

# now select our ligand 
lig = u.select_atoms(select_ligand)

r = RMSD(lig)
results = rmsd.run()

# investigate results
```

See the [userguide](https://userguide.mdanalysis.org/stable/examples/analysis/alignment_and_rms/aligning_trajectory_to_frame.html) and [this](https://docs.mdanalysis.org/2.7.0/documentation_pages/analysis/rms.html) docs page for more info

**Reply by orbeckst** [reply]
Minor comment: you can write the selection 
```python
ligand_selection2 = u2.select_atoms("resname ROH or resname 4GA or resname 2LB or resname 0FA")
```
more compactly as
```python
ligand_selection2 = u2.select_atoms("resname ROH 4GA 2LB 0FA")
```

Btw, I couldn't access the GLYCAM nomenclature link but I found it in the internet archive (http://web.archive.org/web/20230330152511/https://glycam.org/docs/forcefield/glycam-naming-2/index.html). It would be an interesting project to collect all GLYCAM residue names so that we can create a selector "sugar" (or "carbohydrate") that selects anything with those residue names (similar to how "protein" selects anything that has the 20+ amino acid residue names).

**Reply by hmacdope** [reply]
I'm going to raise that as an issue @orbeckst

**Reply by orbeckst** [reply]
I think the solution `rmsd_for_atomgroups2()` in https://github.com/MDAnalysis/mdanalysis/discussions/4562#discussioncomment-9057407 already does the right thing: superimpose on the `reference` (in the example, the protein backbone) and compute RMSD for different groups. (I just would _never_ use `np.round()` on any data and _only_ round when presenting data in a table... but that's just a small criticism of the otherwise nice [talktorial on analyzing MD simulations](https://projects.volkamerlab.org/teachopencadd/talktorials/T020_md_analysis.html)).

The question

> but it would be also nice to see the overall RMSD of the ligand

could be addressed with the selection that @hmacdope suggests in https://github.com/MDAnalysis/mdanalysis/discussions/4562#discussioncomment-9091355 by also **selecting the whole sugar** (here written as `sugar` and as the selection "resname 4GA 2LB 0FA")
```python
ligand_name ='4GA'
ligand_name2='2LB'
ligand_name3='0FA'

# resnames for the whole carbohydrate
sugar = f"{ligand_name} {ligand_name2} {ligand_name3}"

rmsd = rmsd_for_atomgroups2(u, "backbone", ["protein", f"resname {sugar}", f"resname {ligand_name}",f"resname {ligand_name2}", f"resname {ligand_name3}"])
rmsd.plot(title="RMSD of protein and ligand")
plt.ylabel("RMSD (Å)");
```
This should add a RMSD calculation for the whole carbohydrate selection.

**Reply by Noah603** [reply]
Thank you both for the answers and the discussion!  

> Btw, I couldn't access the GLYCAM

Yeha same here, the GLYCAM server seems to be down. Can not access the side at all. You can see the nomencalture as well in the amber tutorial [here on page 48](https://ambermd.org/doc12/Amber22.pdf).

> It would be an interesting project to collect all GLYCAM residue names so that we can create a selector "sugar" (or "carbohydrate") that selects anything with those residue names (similar to how "protein" selects anything that has the 20+ amino acid residue names).

Thats what I was searching for! I think this could be very nice, if there is a simple implementation like that into MDanalysis. 

> ligand_name ='4GA'
ligand_name2='2LB'
ligand_name3='0FA'
resnames for the whole carbohydrate
sugar = f"{ligand_name} {ligand_name2} {ligand_name3}"

But this is also very nice way of doing it, since you can tune the residues you want to see in the analysis. If you have for example a hexasaccharide in which you want to focus during the analysis on only 3 residues. I am going to use that :)

> just would never use np.round() on any data and only round when presenting data in a table.

Good spot! Gonna have that in mind. I really like that talktorial 19 and 20, because it is a great starting point for someone new to the world of MD sims. I was searching for a notebook where the intention behind it was: This is how I anaylsis my MD Protein-Ligand sims, that creates the most important graphs for you, you just need to put your file into it. For example if you get a new student you can just hand it over and get good graphs out.

**Reply by Noah603** [reply]
Hey guys!
I ran into another problem which I wanted to make you aware, incase you want to implement the Glycom nomenclature.
Imagine the sugars sequence is sequence: ROH 4GA VLB 4YB 0LB 4YB 0LB.
Now the above solutions do not work, since we select for the names, but we have twice the same name. In my case this lead to the effect that only one of them gets selected. 
I worked around by renaming them in the prmtop name. This is not a particular nice soultion, but works. 
Best, Noah

**Reply by orbeckst** [reply]
There may be some misunderstanding of how the input for the `RMSD` class works. Let me explain a bit.

## Selections with duplicate resnames

Duplicated resnames should not matter in a normal selection. We just look for anything that has the correct resname:

If you do
```python
sugar_1 = u.select_atoms("resname ROH 4GA VLB 4YB 0LB 4YB 0LB")
sugar_2 = u.select_atoms("resname ROH 4GA 4YB 0LB")
```
then you should get the same atoms selected
```python
sugar_1 == sugar_2
```
should be `True`.

The `sugar_1` atomgroup will contain all the residues that match the name in numerical order, so e.g.
```python
print(list(sugar_1.residues))
```
should show the glyco-residues in order.

## Selecting specific residues with the same name

If you have two sugar residues named *0LB* then you need additional information to tell them apart, for instances, their residue ids.

Lets assume you have a sugar with `ROH-4GA-VLB-4YB-0LB-4YB-0LB` where the resids are
```python
>>> s = u.select_atoms("resname 0LB")
>>> print(s.residues.resids)
[1 2 3 4 5 6 7]
```
(i.e., ROH is resid 1 and the last 0LB is resid 7).

If you want a specific sugar residue such as the second 4YB then you can do a selection such as 
```python
s4YB_2 = u.select_atoms("resid 6 and resname 4YB")
```
(You could probably just use "resid 6" but in some files, the resids are not unique and then you're safer to _also_ specify the resname.)

## Selecting individual residues with `RMSD`

Instead of renaming the residues in your file, you should use more precise selection strings:
```python
rms.RMSD(u, u, select="backbone", 
         groupselections=["protein", "resid 1 and resname ROH", 
                          "resid 4 and resname 4YB", "resid 6 and resname 4YB")
```
Here you would calculate the RMSD for protein, ROH, and both 4YB separately.

If you want to automate generating `groupselections`, something like the following may work
```python
sugar = u.select_atoms("resname ROH 4GA VLB 4YB 0LB")
resnames = sugar.residues.resnames
resids = sugar.residues.resids
groupselections = [f"resid {resid} and resname {resname}" for resid, resname in zip(resids, resnames))]
groupselections.prepend("protein")
```

**Comment by hmacdope** [comment]
@Noah603 could you mark as resolved if your question was answered?

**Comment by orbeckst** [comment]
@Noah603 instead of closing the discussion could you please let us know if your problem was solved (then click the "Mark as answer") or say what's not working.

I am re-opening the discussion so that it remains visible. Closing pretty much removes it from sight, which makes it hard for others to learn from it. This is a bit of a shortcoming in GH Discussions.

**Reply by Noah603** [reply]
ah sorry! mb
